File: services/auth_service/api_key_service.py
"""
API Key management service for Quantum Hub.
"""
from datetime import datetime, timedelta
import secrets
import string
from typing import List, Optional
from bson import ObjectId

# Use direct imports to avoid circular dependency
from models import APIKey, APIKeyCreate, APIKeyResponse, APIUsageStats

# In-memory storage for API keys
api_keys_db = {}

# In-memory storage for API usage stats
api_usage_db = {}

def generate_api_key_string(prefix: str = "qh") -> str:
    """
    Generate a secure API key string.
    
    Args:
        prefix (str): The prefix to use for the API key.
        
    Returns:
        str: A secure random API key string.
    """
    # Generate a random string for the API key
    key_length = 32
    random_part = ''.join(secrets.choice(string.ascii_letters + string.digits) for _ in range(key_length))
    
    # Create the key with a prefix format similar to "qh_pk_randomstring"
    return f"{prefix}_{random_part}"

async def create_api_key(user_id: str, key_data: APIKeyCreate) -> APIKey:
    """
    Create a new API key for a user.
    
    Args:
        user_id (str): The ID of the user.
        key_data (APIKeyCreate): The API key data.
        
    Returns:
        APIKey: The created API key.
    """
    # Generate a new API key
    key_id = str(ObjectId())
    
    # Determine key prefix based on name
    key_prefix = "qh"
    if "prod" in key_data.name.lower() or "production" in key_data.name.lower():
        key_prefix = "qh_pk"
    elif "dev" in key_data.name.lower() or "development" in key_data.name.lower():
        key_prefix = "qh_dk"
    elif "test" in key_data.name.lower() or "trial" in key_data.name.lower():
        key_prefix = "qh_tk"
    
    api_key = APIKey(
        id=key_id,
        key=generate_api_key_string(key_prefix),
        name=key_data.name,
        user_id=user_id,
        created_at=datetime.utcnow(),
        expires_at=key_data.expires_at,
        is_active=True
    )
    
    # Store in in-memory database
    if user_id not in api_keys_db:
        api_keys_db[user_id] = []
    api_keys_db[user_id].append(api_key.dict())
    
    # Initialize usage stats for this key
    api_usage_db[api_key.key] = {
        "request_count": 0,
        "compute_time": 0.0,  # in hours
    }
    
    return api_key

async def get_api_keys(user_id: str) -> List[APIKeyResponse]:
    """
    Get all API keys for a user.
    
    Args:
        user_id (str): The ID of the user.
        
    Returns:
        List[APIKeyResponse]: A list of all API keys for the user.
    """
    if user_id not in api_keys_db:
        return []
    
    api_keys = []
    for key_data in api_keys_db[user_id]:
        key = APIKey(**key_data)
        
        # Mask the key except when it's newly created
        masked_key = f"{key.key[:8]}{'*' * 12}{key.key[-4:]}"
        
        api_keys.append(APIKeyResponse(
            id=key.id,
            key=masked_key,
            name=key.name,
            created_at=key.created_at,
            expires_at=key.expires_at,
            is_active=key.is_active
        ))
    
    return api_keys

async def get_api_key(user_id: str, key_id: str) -> Optional[APIKeyResponse]:
    """
    Get a specific API key for a user.
    
    Args:
        user_id (str): The ID of the user.
        key_id (str): The ID of the API key.
        
    Returns:
        Optional[APIKeyResponse]: The API key if found, None otherwise.
    """
    if user_id not in api_keys_db:
        return None
    
    for key_data in api_keys_db[user_id]:
        if key_data["id"] == key_id:
            key = APIKey(**key_data)
            
            # Mask the key except when it's newly created
            masked_key = f"{key.key[:8]}{'*' * 12}{key.key[-4:]}"
            
            return APIKeyResponse(
                id=key.id,
                key=masked_key,
                name=key.name,
                created_at=key.created_at,
                expires_at=key.expires_at,
                is_active=key.is_active
            )
    
    return None

async def revoke_api_key(user_id: str, key_id: str) -> bool:
    """
    Revoke an API key.
    
    Args:
        user_id (str): The ID of the user.
        key_id (str): The ID of the API key to revoke.
        
    Returns:
        bool: True if the key was revoked, False otherwise.
    """
    if user_id not in api_keys_db:
        return False
    
    for i, key_data in enumerate(api_keys_db[user_id]):
        if key_data["id"] == key_id:
            # Set the key to inactive
            api_keys_db[user_id][i]["is_active"] = False
            return True
    
    return False

async def validate_api_key(api_key: str) -> Optional[str]:
    """
    Validate an API key and return the user ID if it's valid.
    
    Args:
        api_key (str): The API key to validate.
        
    Returns:
        Optional[str]: The user ID if the key is valid, None otherwise.
    """
    for user_id, keys in api_keys_db.items():
        for key_data in keys:
            if key_data["key"] == api_key and key_data["is_active"]:
                # Check if expired
                if key_data["expires_at"] and datetime.utcnow() > key_data["expires_at"]:
                    return None
                
                # Count this request
                if api_key in api_usage_db:
                    api_usage_db[api_key]["request_count"] += 1
                
                return user_id
    
    return None

async def get_api_usage_stats(user_id: str) -> APIUsageStats:
    """
    Get API usage statistics for a user.
    
    Args:
        user_id (str): The ID of the user.
        
    Returns:
        APIUsageStats: The API usage statistics.
    """
    # In a real implementation, these would be retrieved from a database or analytics service
    
    # Default limits
    total_requests_limit = 50000
    compute_time_limit = 100  # hours
    
    # Count totals for this user
    total_requests = 0
    compute_time = 0.0
    
    if user_id in api_keys_db:
        for key_data in api_keys_db[user_id]:
            key = key_data["key"]
            if key in api_usage_db:
                total_requests += api_usage_db[key]["request_count"]
                compute_time += api_usage_db[key]["compute_time"]
    
    # For demo, make sure there's some usage to display
    if total_requests == 0:
        total_requests = 25421
    if compute_time == 0:
        compute_time = 28.5
    
    return APIUsageStats(
        total_requests=total_requests,
        total_requests_limit=total_requests_limit,
        compute_time_hours=compute_time,
        compute_time_limit=compute_time_limit
    )

# Sample data for development
def init_sample_data():
    """Initialize sample API keys for development."""
    # Create some sample data if none exists
    user_id = "sample_user"
    api_keys_db[user_id] = []
    
    # Production key
    prod_key = APIKey(
        id=str(ObjectId()),
        key="qh_pk_sampleproductionkey1234567890abcdefcRtG",
        name="Production Key",
        user_id=user_id,
        created_at=datetime.utcnow() - timedelta(days=90),
        expires_at=None,
        is_active=True
    )
    
    # Development key
    dev_key = APIKey(
        id=str(ObjectId()),
        key="qh_dk_sampledevelopmentkey1234567890abcdeftHj5",
        name="Development Key",
        user_id=user_id,
        created_at=datetime.utcnow() - timedelta(days=45),
        expires_at=datetime.utcnow() + timedelta(days=320),
        is_active=True
    )
    
    # Trial key (expired)
    trial_key = APIKey(
        id=str(ObjectId()),
        key="qh_tk_sampletrialkey1234567890abcdefghijzP4s",
        name="Trial Key",
        user_id=user_id,
        created_at=datetime.utcnow() - timedelta(days=30),
        expires_at=datetime.utcnow() - timedelta(days=1),
        is_active=True
    )
    
    api_keys_db[user_id] = [
        prod_key.dict(),
        dev_key.dict(),
        trial_key.dict()
    ]
    
    # Add some usage data
    api_usage_db[prod_key.key] = {
        "request_count": 15000,
        "compute_time": 18.5,
    }
    api_usage_db[dev_key.key] = {
        "request_count": 9500,
        "compute_time": 8.3,
    }
    api_usage_db[trial_key.key] = {
        "request_count": 921,
        "compute_time": 1.7,
    }

File: frontend/src/app/services/apiKeyService.ts
/**
 * Service for managing API keys
 */
import axios from 'axios';
import { getSession } from 'next-auth/react';

const API_URL = process.env.NEXT_PUBLIC_AUTH_SERVICE_URL || 'http://localhost:8001/api/v1';

export interface ApiKey {
  id: string;
  key: string;
  name: string;
  created_at: string;
  expires_at: string | null;
  is_active: boolean;
}

export interface ApiKeyCreate {
  name: string;
  expires_at?: string | null;
}

export interface ApiUsageStats {
  total_requests: number;
  total_requests_limit: number;
  compute_time_hours: number;
  compute_time_limit: number;
}

// Default mock API keys for demo
const mockApiKeys: ApiKey[] = [
  {
    id: 'mock-key-1',
    key: 'qh_pk_sampleproductionkey123cRtG',
    name: 'Production Key',
    created_at: new Date(Date.now() - 90 * 24 * 60 * 60 * 1000).toISOString(),
    expires_at: null,
    is_active: true
  },
  {
    id: 'mock-key-2',
    key: 'qh_dk_sampledevelopmentkey1tHj5',
    name: 'Development Key',
    created_at: new Date(Date.now() - 45 * 24 * 60 * 60 * 1000).toISOString(),
    expires_at: new Date(Date.now() + 320 * 24 * 60 * 60 * 1000).toISOString(),
    is_active: true
  },
  {
    id: 'mock-key-3',
    key: 'qh_tk_sampletrialkey12345zP4s',
    name: 'Trial Key',
    created_at: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString(),
    expires_at: new Date(Date.now() - 1 * 24 * 60 * 60 * 1000).toISOString(),
    is_active: false
  }
];

// In-memory store for the demo
let demoApiKeys = [...mockApiKeys];

/**
 * Get authentication headers for API requests
 */
const getAuthHeaders = async () => {
  try {
    const session = await getSession();
    if (session?.accessToken) {
      return {
        'Authorization': `Bearer ${session.accessToken}`,
        'Content-Type': 'application/json',
      };
    }

    // Fallback for demo
    return {
      'Content-Type': 'application/json',
    };
  } catch (error) {
    console.error('Error getting auth headers:', error);
    // Fallback for demo
    return {
      'Content-Type': 'application/json',
    };
  }
};

/**
 * Get all API keys for the current user
 */
export const getApiKeys = async (): Promise<ApiKey[]> => {
  try {
    const headers = await getAuthHeaders();
    console.log('Fetching API keys from:', `${API_URL}/api-keys`);
    const response = await axios.get(`${API_URL}/api-keys`, {
      headers,
      withCredentials: true,
    });

    // The response is now directly the array of API keys
    return response.data;
  } catch (error) {
    console.error('Error fetching API keys from server:', error);

    // Return the in-memory demo data
    return demoApiKeys;
  }
};

/**
 * Create a new API key
 */
export const createApiKey = async (apiKeyData: ApiKeyCreate): Promise<ApiKey> => {
  try {
    const headers = await getAuthHeaders();
    console.log('Creating API key at:', `${API_URL}/api-keys`);
    const response = await axios.post(`${API_URL}/api-keys`, apiKeyData, {
      headers,
      withCredentials: true,
    });

    // The response is now directly the created API key
    return response.data;
  } catch (error) {
    console.error('Error creating API key:', error);

    // Generate a mock key for demo
    const keyPrefix = apiKeyData.name.toLowerCase().includes('prod') ? 'qh_pk' :
                    apiKeyData.name.toLowerCase().includes('dev') ? 'qh_dk' : 'qh_tk';

    // Generate a random suffix for the key
    const randomSuffix = Math.random().toString(36).substring(2, 10);

    const newKey: ApiKey = {
      id: `mock-key-${Date.now()}`,
      key: `${keyPrefix}_${randomSuffix}`,
      name: apiKeyData.name,
      created_at: new Date().toISOString(),
      expires_at: apiKeyData.expires_at || null,
      is_active: true
    };

    // Add to our in-memory demo data
    demoApiKeys = [newKey, ...demoApiKeys];

    return newKey;
  }
};

/**
 * Revoke (disable) an API key
 */
export const revokeApiKey = async (keyId: string): Promise<void> => {
  try {
    const headers = await getAuthHeaders();
    await axios.delete(`${API_URL}/api-keys/${keyId}`, {
      headers,
      withCredentials: true,
    });
  } catch (error) {
    console.error('Error revoking API key:', error);

    // For demo, update our in-memory data
    demoApiKeys = demoApiKeys.map(key =>
      key.id === keyId ? { ...key, is_active: false } : key
    );
  }
};

/**
 * Get API usage statistics
 */
export const getApiUsageStats = async (): Promise<ApiUsageStats> => {
  try {
    const headers = await getAuthHeaders();
    console.log('Fetching API usage stats from:', `${API_URL}/api-keys/usage/stats`);
    const response = await axios.get(`${API_URL}/api-keys/usage/stats`, {
      headers,
      withCredentials: true,
    });

    // The response is now directly the API usage stats
    return response.data;
  } catch (error) {
    console.error('Error fetching API usage stats:', error);

    // Return demo stats
    return {
      total_requests: 25421,
      total_requests_limit: 50000,
      compute_time_hours: 28.5,
      compute_time_limit: 100
    };
  }
};
